server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentAction.java

```diff
@@ -0,0 +1,251 @@
+    writeProtobuf(componentWsResponse, request, response);
+  }
+
+  private ComponentWsResponse doHandle(ComponentWsRequest request) {
+    DbSession dbSession = dbClient.openSession(false);
+    try {
+      ComponentDto component = componentFinder.getByUuidOrKey(dbSession, request.getComponentId(), request.getComponentKey(), COMPONENT_ID_AND_KEY);
+      Optional<ComponentDto> refComponent = getReferenceComponent(dbSession, component);
+      checkPermissions(component);
+      SnapshotDto lastSnapshot = dbClient.snapshotDao().selectLastSnapshotByComponentId(dbSession, component.getId());
+      List<MetricDto> metrics = searchMetrics(dbSession, request);
+      List<WsMeasures.Period> periods = snapshotsToPeriods(lastSnapshot);
+      List<MeasureDto> measures = searchMeasures(dbSession, component, lastSnapshot, metrics, periods);
+
+      return buildResponse(request, component, refComponent, measures, metrics, periods);
+    } finally {
+      dbClient.closeSession(dbSession);
+    }
+  }
+
+  private Optional<ComponentDto> getReferenceComponent(DbSession dbSession, ComponentDto component) {
+    if (component.getCopyResourceId() == null) {
+      return Optional.absent();
+    }
+
+    return dbClient.componentDao().selectById(dbSession, component.getCopyResourceId());
+  }
+
+  private ComponentWsResponse buildResponse(ComponentWsRequest request, ComponentDto component, Optional<ComponentDto> refComponent, List<MeasureDto> measures,
+    List<MetricDto> metrics, List<WsMeasures.Period> periods) {
+    ComponentWsResponse.Builder response = ComponentWsResponse.newBuilder();
+    Map<Integer, MetricDto> metricsById = Maps.uniqueIndex(metrics, MetricDtoFunctions.toId());
+    Map<MetricDto, MeasureDto> measuresByMetric = new HashMap<>();
+    for (MeasureDto measure : measures) {
+      MetricDto metric = metricsById.get(measure.getMetricId());
+      measuresByMetric.put(metric, measure);
+    }
+    Map<Long, String> referenceComponentUuidById = new HashMap<>();
+    if (refComponent.isPresent()) {
+      referenceComponentUuidById.put(refComponent.get().getId(), refComponent.get().uuid());
+    }
+
+    response.setComponent(componentDtoToWsComponent(component, measuresByMetric, referenceComponentUuidById));
+
+    List<String> additionalFields = request.getAdditionalFields();
+    if (additionalFields != null) {
+      if (additionalFields.contains(ADDITIONAL_METRICS)) {
+        for (MetricDto metric : metrics) {
+          response.getMetricsBuilder().addMetrics(metricDtoToWsMetric(metric));
+        }
+      }
+      if (additionalFields.contains(ADDITIONAL_PERIODS)) {
+        response.getPeriodsBuilder().addAllPeriods(periods);
+      }
+    }
+
+    return response.build();
+  }
+
+  private List<MetricDto> searchMetrics(DbSession dbSession, ComponentWsRequest request) {
+    List<MetricDto> metrics = dbClient.metricDao().selectByKeys(dbSession, request.getMetricKeys());
+    if (metrics.size() < request.getMetricKeys().size()) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'request' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)